### DISCUSSION TITLE
Resetting a ZedGraph plot

### DISCUSSION THREAD

**Discussion by ZegCricket** [bonsai-rx/bonsai#2085] [Q&A]
Hello! I hope you are doing well.

I'm starting to use the GUI package for my project and I need to plot some data in real-time, but I need to reset the plots at the beginning of every block without having to close and reopen the GUI visualizer. Is there currently a way to do that?

Thank you in advance for the help!

**Comment by banchan86** [comment]
As far as I know, there's no easier way in the editor (say a keyboard shortcut) that would reset the visualizer, but there is a way to achieve that in Bonsai workflows like in the first example [here](https://github.com/bonsai-rx/bonsai/issues/2075). Does that work for you?

**Reply by ZegCricket** [reply]
Thank you for the reply! I hadn't thought about this solution, but unfortunately it doesn't work for me, because I would like to integrate the plot in a TableLayoutPanel, which doesn't seem to work just yet.

**Comment by ZegCricket** [comment]
For the majority of my plots (although not for every single one), I was planning on just use the `Capacity` and/or `Span` properties, but they don't seem to work for the GraphPanel node. I used this [example](https://fchampalimaud.github.io/cf.bonsai/workflows/BonsaiExamples/GUI/LineGraph/LineGraph.html) as a starting point to test this.

**Reply by banchan86** [reply]
Hi ZegCricket! Yea an issue has been raised for this https://github.com/bonsai-rx/gui/issues/33, you can follow that issue for any updates.

**Reply by ZegCricket** [reply]
Thank you, I will! :)

It would be cool that the Gui.ZedGraph package would be further developed to accomodate some common needs (plot resetting, title, axis labels) which if exist are not super evident. Because if I want to run a workflow in no-editor mode, it's a bit inconvenient that the current solution is to close and reopen the visualizer and that may cause me to lose some data I wanted to keep in my plots.

**Reply by bruno-f-cruz** [reply]
I think that there are already solutions for the features you are requesting. This doesnt mean they cant be improved obviously, but I would encourage you to play a bit more with the package and see how far you get! If you find these useful, consider making a pull-request documenting them to the Bonsai.Gui package repository. Thanks!
1. Plot resetting, does this work? The plots already seem to flush if the sequence resubscribes.
![image](https://github.com/user-attachments/assets/ceaf3dba-a0a2-46c9-9a4a-0e03cc8f4cf0)
2. For titles, you can use a TableLayoutPanel title (You can get extra fancy and even make a title out of a textbox, button or label):
![image](https://github.com/user-attachments/assets/95d583b4-d486-45cd-9a7a-f4196b983851)
Inside the GroupWorkflow:
![image](https://github.com/user-attachments/assets/b240c84c-235a-490b-883c-c9bae45bafb5)
3. For Axis you can use `ExpressionTransforms`:
```
new(it.Value as Label, it.Timestamp as Timestamp)
```
![image](https://github.com/user-attachments/assets/4af819da-4bf9-4a73-96a2-9a0a2f967619)

**Reply by ZegCricket** [reply]
Thank you for the reply! When you said pull-request, I'm assuming you meant to open an issue (I think I probably will at some point).

1. The example in specific works, but if I want to propagate that reset of the plot to a complete GUI (with buttons, labels, other plots) I don't know how, because you would have to branch out the take and repeat part, since the `Repeat` node doesn't have the plot visualizer, which doesn't seem to work.
2. That's a nice suggestion, thanks!
3. For some reason, I can't make this example to work. It's simple and it seems that you got it working, so I don't know what's happening :/

**Reply by bruno-f-cruz** [reply]
I meant a pull-request where one could add more examples to documentation of the package

1. Cant you do the same as I did in the second example and use the repeat to resubscribe to the full UI? Can you attach a minimal example?
For 3, can you try this?

```xml
<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.5"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:scr="clr-namespace:Bonsai.Scripting.Expressions;assembly=Bonsai.Scripting.Expressions"
                 xmlns:zg="clr-namespace:Bonsai.Gui.ZedGraph;assembly=Bonsai.Gui.ZedGraph"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timer">
          <rx:DueTime>PT0S</rx:DueTime>
          <rx:Period>PT1S</rx:Period>
        </Combinator>
      </Expression>
      <Expression xsi:type="scr:ExpressionTransform">
        <scr:Expression>new(it as Foo, it as Bar)</scr:Expression>
      </Expression>
      <Expression xsi:type="zg:RollingGraphBuilder">
        <zg:IndexSelector>Bar</zg:IndexSelector>
        <zg:ValueSelector>Foo</zg:ValueSelector>
        <zg:SymbolType>None</zg:SymbolType>
        <zg:LineWidth>1</zg:LineWidth>
        <zg:CurveSettings />
        <zg:Capacity xsi:nil="true" />
        <zg:Min xsi:nil="true" />
        <zg:Max xsi:nil="true" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="1" To="2" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>
```

**Reply by ZegCricket** [reply]
Got it! I'll try to do it at some point.

1. You can, but I think it only works for that case where you want to reset the full UI. If, for example, you have a UI with 2 plots and you only want to reset one of the them and preserve the other, I think you're going to face the issue I was describing.

For 3, it worked, thanks! I forgot to configure the `IndexSelector` and the `ValueSelector` properties.

**Reply by bruno-f-cruz** [reply]
Yes in that case I don't see a solution with the current package. Def something to consider for future versions!

**Reply by ZegCricket** [reply]
Yes, it would be very cool!

And thank you so much for all of this back-and-forth conversation, it was really helpful, I think I'll use some of the tips you gave me in my project! :)
